### DISCUSSION TITLE
[New device support]: Linptech/Tuya self powered switch (GreenPower) apparently returns random "commandID"

### DISCUSSION THREAD

**Discussion by uizin** [Koenkk/zigbee2mqtt#19352] [Devices]
Hello, I'm a noob but am working hard to understand how to code a converter for a new device I got.
It is a Two-rocker switch, self powered with no battery (cynetic), and works on the Zigbee Green Power Protocol.
It is advertised as Tuya, nut the device I got is branded linptech, and the model is K9Z.

I managed to pair it to my zigbee2mqtt network, and I am identifying it using an external converter.

The problem is that the debug log messages show a different "commandID" value each time one of the buttons is pressed. No matter which button I press, there is not a consistent value for the right and left button presses. The ID does not seem to follow a specific pattern, and I do not find a way to differentiate right button and left button presses.

Here the examples from the log. Only the "commandID" value changes on each press.

```
Received Zigbee message from '0x00...', type 'commandNotification', cluster 'greenPower', data '{"commandID":99,"frameCounter":249,"srcID":93102668}' from endpoint 242 with groupID null
```
```
Received Zigbee message from '0x00...', type 'commandNotification', cluster 'greenPower', data '{"commandID":121,"frameCounter":2,"srcID":93102668}' from endpoint 242 with groupID null
```

What am I doing wrong?
Here the full external converter I am using to integrate the device:

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const utils = require('zigbee-herdsman-converters/lib/utils');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['GreenPower_2'],
    model: 'K9Z',
    vendor: 'linptech',
    description: 'Self-powered Zigbee Wireless Switch',
    fromZigbee: [{
      cluster: "greenPower",
      type: ["commandNotification"],
      convert: (model, msg, publish, options, meta) => {
        const commandID = msg.data.commandID;
        const lookup = {0x34: "left_btn", 0x35: "right_btn"};
        if(utils.hasAlreadyProcessedMessage(msg, model, msg.data.frameCounter, `${msg.device.ieeeAddr}_${commandID}`)) {
          meta.logger.debug(`GreenPower_2 command '${commandID}' already processed`);
          return;
        }
        if(commandID === 224) return;
        if(!lookup.hasOwnProperty(commandID)) {
          meta.logger.error(`GreenPower_2 error: missing command '${commandID}'`);
        }else{
          return {action: lookup[commandID]};
        }
      }
    }],
    toZigbee: [],
    exposes: [e.action(["left_btn", "right_btn"])], // Defines what this device exposes, used for e.g. Home Assistant discovery and >
};

module.exports = definition;
```

Thank you for any suggestion!!

**Comment by felix63** [comment]
Any progess made? I have a similar one-button device. I see:
`Zigbee2MQTT:debug 2024-01-31 10:28:23: Received Zigbee message from 'Button 9', type 'commandCommissioningNotification', cluster 'greenPower', data '{"commandFrame":{"raw":{"data":[196,33,208,33,221,67,72],"type":"Buffer"}},"commandID":159,"frameCounter":991,"gppGddLink":208,"gppNwkAddr":8644,"options":2864,"payloadSize":0,"srcID":356401669}' from endpoint 242 with groupID 0`

**Reply by uizin** [reply]
Still stuck on this... did not find a solution

**Comment by dduransseau** [comment]
My understanding of this issue is that Green Power endpoint powered by Tuya support a higher level of encryption. then what it is commonly use.
During commissioning the extended options is `243` for tuya switch but `242` for a FoH switch. Wich imply `0b11 (Encryption & full (4B) frame counter and full (4B) MIC)` compared to `0b10 (Full (4B) frame counter and full (4B) MIC only)` as security level, following p43-44 of [specification](https://zigbeealliance.org/wp-content/uploads/2019/11/docs-09-5499-26-batt-zigbee-green-power-specification.pdf)
Right now it seems that encryption is not supported by z2m, I had a look at the Green Power implementation but I don't really know how this should be implemented.

**Reply by ambroseus** [reply]
found this information (also try to integrate tuya self-powered switch):

https://zigbee.blakadder.com/Illumra_ZBT-S2AWH.html

Green Power is a feature of Zigbee that allows for energy-harvesting technology to be used directly with the Zigbee stack. Green Power devices do not send standard Zigbee messages. To use a Green Power device in your Zigbee network you need a "translator". Currently only a Philips Hue device can serve as a translator.
Your Zigbee Green Power device has to always be in range of a Philips Hue device to function

**Comment by tshakah** [comment]
I think solving this would also solve https://github.com/Koenkk/zigbee2mqtt/issues/19405. I tried this converter with that switch and was getting the same result

**Comment by Lakitna** [comment]
I also got one of these switches, but I can't see any command IDs at all. Since I have the 1-gang version, the changing command IDs shouldn't be an issue for me, if they actually show up :)

Some debug questions:

- Configuring the device works without any issues. It shows up in Z2M as expected with the description from your external definition.
- I have a Hue device close for the weird translation stuff. 
- I added a log line at the first line of the `convert` function. Nothing ever gets logged.
- I also added a `configure` function. Stuff is logged here as expected.

It feels like it does absolutely nothing with anything in the `fromZigbee` array.

I did notice that there is a second way of defining devices with the `modernExtends` stuff. Can this be related? Is it worth rebuilding this with modern extends?

**Reply by dduransseau** [reply]
This switch should work on dev branch, following [#19405](https://github.com/Koenkk/zigbee2mqtt/issues/19405)

**Reply by chris-1243** [reply]
https://github.com/Koenkk/zigbee-herdsman/pull/1363

There has been tons of work done this month to clean ZGP stuff and add support to new device.

Latest-dev should normally on the right track to cover all of this👍

**Comment by mrbluebrett** [comment]
Hello, I have the same switch but my zigbee2mqtt does not see anything when trying to pair...
I set my zigbee channel (20) on the 9 position of the selector.
I have a TRETAKT plug in range which is supposed to act as a green power proxy.
I am missing something to activate the proxy function?
I can't see anything even in the logs when trying to pair...

**Reply by chris-1243** [reply]
Lots of improvements have been done this month on ZigeeGreenPower devices and routers acting as a proxy like your Ikea Tretakt.

Either you could try latest-dev or I would wait the april release in which all those corrections will be available. Thanks @Nerivec for all the work done on this.

**Reply by mrbluebrett** [reply]
I have read that in the next release or current dev branch this should be fully supported, but before that, people where seeing the device or paring it even if it does not work properly. That's why I'm asking if i should do something else to pair or at least see something in the logs...

**Reply by chris-1243** [reply]
What coordinator do you have?
TI chip, it would have been quite complicated before the update of the code.
Silabs chip, it was possible with lots of drawback as described in this discussion

**Reply by mrbluebrett** [reply]
Sonoff USB dongle-P. I think it's TI, right?

**Reply by chris-1243** [reply]
Indeed, TI chip. I would wait for the april release

**Reply by mrbluebrett** [reply]
I have updated z2m to version 2.2 and still nothing appearing for this kinetic switch...

**Comment by ginkage** [comment]
I've updated to 2.2, and my switches seem to work, but — for some reason, they only work when I have "permit join" enabled. As soon as permit join is disabled, I stop getting the commands...

**Reply by dominikmicuta** [reply]
https://github.com/Koenkk/zigbee2mqtt/issues/25620

**Comment by joh-nnyS** [comment]
I recently got two of these switches (linptech K9Z(TY) 2 gang and 1 gang) for testing purposes. Currently on Zigbee2MQTT 2.5.1-1, with a SONOFF Zigbee 3.0 USB Dongle Plus running 20250321 firmware.

I can pair both of them, but they are listed as unsupported. After following: [this](https://github.com/Koenkk/zigbee2mqtt/issues/27182#issuecomment-2821180745) they show up as supported, but I don't seem to see any signals.

[#25620](https://github.com/Koenkk/zigbee2mqtt/issues/25620 ) seems to suggest they may work with a different coordinator? SkyConnect (Ember) not sure if that applies to these switches?

**Comment by M-ax-G** [comment]
I've got 20 of the Linptech single and double switches.
The **channel selector needs to be set to "E"** in order for the switches to work. Pairing is done by pressing the pairing button and pushing the rocker once; then release the pairing button. This **pairing has to be done per rocker switch**. So a double is paired with each of the two rockers.
z2m needs to be in join mode. I am using a SONOFF Zigbee 3.0 USB Dongle Plus | ZBDongle-E. z2m 2.9.1 allows pairing, while 2.14.1 shows them on dashboard and home screens.
